feature/preview/src/main/java/com/google/jetpackcamera/feature/preview/PreviewViewModel.kt

```diff
@@ -110,14 +116,116 @@ class PreviewViewModel @AssistedInject constructor(
         }
     }
 
+    private fun getCaptureToggleUiState(
+        systemConstraints: SystemConstraints,
+        cameraAppSettings: CameraAppSettings
+    ): CaptureModeToggleUiState {
+        val cameraConstraints: CameraConstraints? = systemConstraints.forCurrentLens(
+            cameraAppSettings
+        )
+        val hdrDynamicRangeSupported = cameraConstraints?.let {
+            it.supportedDynamicRanges.size > 1
+        } ?: false
+        val hdrImageFormatSupported = cameraConstraints?.let {
+            it.supportedImageFormatsMap[cameraAppSettings.captureMode]!!.size > 1
+        } ?: false
+        val isShown = previewMode is PreviewMode.ExternalImageCaptureMode ||
+            cameraAppSettings.imageFormat == ImageOutputFormat.JPEG_ULTRA_HDR ||
+            cameraAppSettings.dynamicRange == DynamicRange.HLG10
+        val enabled = previewMode !is PreviewMode.ExternalImageCaptureMode &&
+            hdrDynamicRangeSupported && hdrImageFormatSupported
+        return if (isShown) {
+            val currentMode = if (previewMode is PreviewMode.ExternalImageCaptureMode ||
+                cameraAppSettings.imageFormat == ImageOutputFormat.JPEG_ULTRA_HDR
+            ) {
+                CaptureModeToggleUiState.ToggleMode.CAPTURE_TOGGLE_IMAGE
+            } else {
+                CaptureModeToggleUiState.ToggleMode.CAPTURE_TOGGLE_VIDEO
+            }
+            if (enabled) {
+                CaptureModeToggleUiState.Enabled(currentMode)
+            } else {
+                CaptureModeToggleUiState.Disabled(
+                    currentMode,
+                    getCaptureToggleUiStateDisabledReason(
+                        hdrDynamicRangeSupported,
+                        hdrDynamicRangeSupported,
+                        systemConstraints,
+                        cameraAppSettings.cameraLensFacing,
+                        cameraAppSettings.captureMode
+                    )
+                )
+            }
+        } else {
+            CaptureModeToggleUiState.Invisible
+        }
+    }
+
+    @SuppressLint("RestrictedApi")
+    private fun getCaptureToggleUiStateDisabledReason(
+        hdrDynamicRangeSupported: Boolean,
+        hdrImageFormatSupported: Boolean,
+        systemConstraints: SystemConstraints,
+        currentLensFacing: LensFacing,
+        currentCaptureMode: CaptureMode
+    ): CaptureModeToggleUiState.DisabledReason {
+        if (previewMode is PreviewMode.ExternalImageCaptureMode) {
+            return CaptureModeToggleUiState.DisabledReason.VIDEO_CAPTURE_EXTERNAL_UNSUPPORTED
+        }
+        return if (!hdrImageFormatSupported) {
+            for (lens in LensFacing.values()) {
+                val cameraConstraints = systemConstraints.perLensConstraints[lens]
+                for (captureMode in CaptureMode.values()) {
+                    if (cameraConstraints?.let {
+                            it.supportedImageFormatsMap[captureMode]!!.size > 1
+                        } == true
```

Comment:
You should avoid the `!!` operator (see go/kotlin-practices/null#avoid-not-null-assertion-operator).

You can replace this with:
```
if (cameraConstraints?.supportedImageFormatsMap?.get(captureMode)?.let { it.size > 1 } == true)
```